VERSIONING.md

```diff
@@ -0,0 +1,206 @@
+  differences on stable branches).
+
+### Mantainers
+
+Don't be lazy, read the rest of this doc :-)
+
+## Overview
+
+controller-runtime (and friends) follow [Semantic
+Versioning](https://semver.org).  I'd reccomend reading the aforementioned
+link if you're not familiar, but essentially, for any given release X.Y.Z:
+
+- an X (*major*) release indicates a set of backwards-compatible code.
+  Changing X means there's a breaking change.
+
+- a Y (*minor*) release indicates a minimum feature set.  Changing Y means
+  the addition of a backwards-compatible feature.
+
+- a Z (*patch*) release indicates minimum set of bugfixes.  Changing
+  Z means a backwards-compatible change that doesn't add functionality.
+
+*NB*: If the major release is `0`, any minor release may contain breaking
+changes.
+
+These guarantees extend to all code exposed in public APIs of
+controller-runtime. This includes code both in controller-runtime itself,
+*plus types from dependencies in public APIs*.  Types and functions not in
+public APIs are not considered part of the guarantee.
+
+In order to easily maintain the guarantees, we have a couple of processes
+that we follow.
+
+## Branches
+
+controller-runtime contains two types of branches: the *master* branch and
+*release-X* branches.
+
+The *master* branch is where development happens.  All the latest and
+greatest code, including breaking changes, happens on master.
+
+The *release-X* branches contain stable, backwards compatible code.  Every
+major (X) release, a new such branch is created.  It is from these
+branches that minor and patch releases are tagged.
+
+The maintainers are responsible for updating the contents of this branch;
+generally, this is done just before a release using release tooling that
+filters and checks for changes tagged as breaking (see below).
+
+### Tooling
+
+***TODO***
+
+## PR Process
+
+Every PR should be annotated with an icon indicating whether it's
+a:
+
+- Breaking change: :warning: (`:warning:`)
+- Non-breaking feature: :sparkles: (`:sparkles:`)
+- Patch fix: :bug: (`:bug:`)
+- Docs: :book: (`:book:`)
+- Infra/Tests/Other: :running: (`:running:`)
+
+Individual commits may be tagged separately, but will generally be assumed
+to match the PR. For instance, if you have a bugfix in with a breaking
+change, it's generally encouraged to submit the bugfix separately, but if
+you must put them in one PR, mark the commit separately.
+
+## Release Process
+
+Minor and patch releases are generally done immediately after a feature or
+bugfix is landed, or sometimes a series of features tied together.
+
+Major releases are done once a sufficient amount of breaking changes are
+accrued.  Since we don't intend to have a ton of these, the maintainers
+will evaluate when to do a major release as it comes up.
+
+### Exact Steps
+
+***TODO***
+
+### Breaking Changes
+
+Try to avoid breaking changes.  They make life difficult for users, who
```

Comment:
Can we provide a `sed` operation to automatically fix any breaking changes with defaults that maintain the old behavior or provide some new default?

Reply:
Probably want more intelligent than sed, but it should be possible to do so